### PR TITLE
feat(cli): Stop workflows by label selector and field selector

### PR DESCRIPTION
**Pull request by dinever**
Part of #3931

Screenshot from 2021-04-24 21-48-03

### CODE REVIEW

File: cmd/argo/commands/stop.go [hunk 1]
```diff
@@ -31,28 +46,75 @@ func NewStopCommand() *cobra.Command {
   argo stop @latest
```
Comment (alexec) [inline cmd/argo/commands/stop.go:46]: suggestion - add an example of the new usages

File: cmd/argo/commands/stop.go [hunk 2]
```diff
@@ -31,28 +46,75 @@ func NewStopCommand() *cobra.Command {
...
-				errors.CheckError(err)
-				fmt.Printf("workflow %s stopped\n", wf.Name)
-			}
+			err := stopWorkflows(ctx, serviceClient, stopArgs, args)
+			errors.CheckError(err)
 		},
 	}
 	command.Flags().StringVar(&stopArgs.message, "message", "", "Message to add to previously running nodes")
 	command.Flags().StringVar(&stopArgs.nodeFieldSelector, "node-field-selector", "", "selector of node to stop, eg: --node-field-selector inputs.paramaters.myparam.value=abc")
+	command.Flags().StringVarP(&stopArgs.labelSelector, "selector", "l", "", "Selector (label query) to filter on, supports '=', '==', and '!='.(e.g. -l key1=value1,key2=value2)")
```
Comment (alexec) [inline cmd/argo/commands/stop.go:63]: can I please confirm these descriptions are consistent with other commands?

File: cmd/argo/commands/stop.go [hunk 3]
```diff
@@ -31,28 +46,75 @@ func NewStopCommand() *cobra.Command {
...
+				Name:      n,
+				Namespace: stopArgs.namespace,
+			},
+		})
+	}
+
+	stoppedNames := make(map[string]bool)
+	for _, wf := range wfs {
+		if _, ok := stoppedNames[wf.Name]; ok {
+			// de-duplication in case there is a overlap between the selector and given workflow names
```
Comment (alexec) [inline cmd/argo/commands/stop.go:99]: nice

File: cmd/argo/commands/stop.go [hunk 4]
```diff
@@ -31,28 +46,75 @@ func NewStopCommand() *cobra.Command {
...
+		wf, err := serviceClient.StopWorkflow(ctx, &workflowpkg.WorkflowStopRequest{
+			Name:              wf.Name,
+			Namespace:         wf.Namespace,
+			NodeFieldSelector: selector.String(),
+			Message:           stopArgs.message,
+		})
+		if err != nil {
+			return err
+		}
+		fmt.Printf("workflow %s stopped\n", wf.Name)
```
Comment (alexec) [inline cmd/argo/commands/stop.go:117]: nice - `println` prints to stderr (for some mad reason)

File: cmd/argo/commands/stop_test.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+	workflowpkg "github.com/argoproj/argo-workflows/v3/pkg/apiclient/workflow"
+	workflowmocks "github.com/argoproj/argo-workflows/v3/pkg/apiclient/workflow/mocks"
+	wfv1 "github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1"
+)
+
+func Test_stopWorkflows(t *testing.T) {
+	t.Run("Stop workflow dry-run", func(t *testing.T) {
+		c := &workflowmocks.WorkflowServiceClient{}
+		stopArgs := stopOps{
```
Comment (alexec) [inline cmd/argo/commands/stop_test.go:20]: good stuff

**Review by alexec — APPROVED** [review]

File: cmd/argo/commands/stop.go (hunk 1 above)
Comment (dinever) [reply cmd/argo/commands/stop.go:47]: Thanks! added examples for both `argo stop` and `argo terminate` in the latest commit:

Screen Shot 2021-04-28 at 12 44 39 AM

Screen Shot 2021-04-28 at 12 45 06 AM

File: cmd/argo/commands/stop.go (hunk 2 above)
Comment (dinever) [reply cmd/argo/commands/stop.go]: Thanks! Looks like the label selector descriptions from the existing commands are not quite consistent. Here is a summary:

- `argo stop`: Selector (label query) to filter on, supports '=', '==', and '!='.(e.g. -l key1=value1,key2=value2)
- `argo list`: Selector (label query) to filter on, supports '=', '==', and '!='.(e.g. -l key1=value1,key2=value2)
- `argo delete`: Selector (label query) to filter on, not including uninitialized ones
- `argo terminate`: Selector (label query) to filter on, not including uninitialized ones
- `argo archive list`: Selector (label query) to filter on, not including uninitialized ones

In the latest commit, I've changed all of them to `Selector (label query) to filter on, not including uninitialized ones, supports '=', '==', and '!='.(e.g. -l key1=value1,key2=value2)`

The field selector does have consistent descriptions.

Let me know if you have any feedback on this. Thanks!

**Review by alexec — APPROVED** [review]
Good stuff. I approve of this.

**Pull request merged**